include/deal.II/optimization/solver_bfgs.h

```diff
@@ -0,0 +1,389 @@
+//-----------------------------------------------------------
+//
+//    Copyright (C) 2018 by the deal.II authors
+//
+//    This file is part of the deal.II library.
+//
+//    The deal.II library is free software; you can use it, redistribute
+//    it, and/or modify it under the terms of the GNU Lesser General
+//    Public License as published by the Free Software Foundation; either
+//    version 2.1 of the License, or (at your option) any later version.
+//    The full text of the license can be found in the file LICENSE.md at
+//    the top level directory of deal.II.
+//
+//---------------------------------------------------------------
+
+#ifndef dealii_solver_bfgs_h
+#define dealii_solver_bfgs_h
+
+#include <deal.II/lac/solver.h>
+
+#include <deal.II/numerics/history.h>
+
+#include <deal.II/optimization/line_minimization.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * Implements limited memory BFGS minimization method.
+ *
+ * Hessian is updated using the formula of Broyden, Fletcher, Goldfarb, and
+ * Shanno (BFGS): \f[ H^{(k+1)} &= \left[I -\rho_{(k)} s^{(k)} \otimes y^{(k)}
+ * \right] \gz H^{(k)} \left[I -\rho^{(k)} y^{(k)} \otimes s^{(k)} \right]
+ * +        \rho^{(k)} s^{(k)} \otimes s^{(k)}  \\
+ * y^{(k)} &:= g^{(k+1)} - g^{(k)} \\
+ * s^{(k)} &:= x^{(k+1)} - x^{(k)} \\
+ * \rho^{(k)} &:= \frac{1}{y^{(k)} \cdot s^{(k)}}
+ * \f]
+ * for a symmetric positive definite $H$. Limited memory variant is
+ * implemented via the two-loop recursion.
+ *
+ * @author Denis Davydov, 2018
+ */
+template <typename VectorType>
+class SolverBFGS : public Solver<VectorType>
+{
+public:
+  /**
+   * Number type.
+   */
+  typedef typename VectorType::value_type Number;
+
+
+  /**
+   * Standardized data struct to pipe additional data to the solver.
+   */
+  struct AdditionalData
+  {
+    /**
+     * Constructor.
+     */
+    explicit AdditionalData(const unsigned int max_history_size = 5,
+                            const bool         debug_output     = false);
+
+    /**
+     * Maximum history size.
+     */
+    unsigned int max_history_size;
+
+    /**
+     * Print extra debug output to deallog.
+     */
+    bool debug_output;
+  };
+
+
+  /**
+   * Constructor.
+   */
+  explicit SolverBFGS(SolverControl &       residual_control,
+                      const AdditionalData &data = AdditionalData());
+
+  /**
+   * Virtual destructor.
+   */
+  virtual ~SolverBFGS() = default;
+
+  /**
+   * Solve the problem starting from initial state @p x.
+   *
+   * The function @p compute should update the value of gradient of function (second parameter)
+   * based on the input value of unknown (first parameter) and return
+   * the value of the function being minimized.
+   *
+   * @note The value of unknown is intentionally kept as non-const reference
+   * to allow its modifications when it represent a subspace.
```

Comment:
you are referring to the function, or to the std::function object? In any case, I don't understand what you want to modify and why -- or are you saying that the *result of the minimization* is returned in the second argument?